cert/cert.go

```diff
@@ -0,0 +1,203 @@
+	certPEMData := []byte(certPEM)
+	for len(certPEMData) > 0 {
+		var certBlock *pem.Block
+		certBlock, certPEMData = pem.Decode(certPEMData)
+		if certBlock == nil {
+			break
+		}
+		if certBlock.Type == "CERTIFICATE" {
+			cert, err := x509.ParseCertificate(certBlock.Bytes)
+			return cert, err
+		}
+	}
+	return nil, errors.New("no certificates found")
+}
+
+// ParseCertAndKey parses the given PEM-formatted X509 certificate
+// and RSA private key.
+func ParseCertAndKey(certPEM, keyPEM string) (*x509.Certificate, *rsa.PrivateKey, error) {
+	tlsCert, err := tls.X509KeyPair([]byte(certPEM), []byte(keyPEM))
+	if err != nil {
+		return nil, nil, err
+	}
+
+	cert, err := x509.ParseCertificate(tlsCert.Certificate[0])
+	if err != nil {
+		return nil, nil, err
+	}
+
+	key, ok := tlsCert.PrivateKey.(*rsa.PrivateKey)
+	if !ok {
+		return nil, nil, fmt.Errorf("private key with unexpected type %T", key)
+	}
+	return cert, key, nil
+}
+
+// NewCA generates a CA certificate/key pair suitable for signing server
+// keys for an environment with the given name.
+func NewCA(commonName, UUID string, expiry time.Time) (certPEM, keyPEM string, err error) {
+	key, err := rsa.GenerateKey(rand.Reader, KeyBits)
+	if err != nil {
+		return "", "", err
+	}
+	// TODO(perrito666) 2016-05-02 lp:1558657
+	now := time.Now()
+
+	serialNumber, err := newSerialNumber()
+	if err != nil {
+		return "", "", errors.Trace(err)
+	}
+
+	template := &x509.Certificate{
+		SerialNumber: serialNumber,
+		Subject: pkix.Name{
+			CommonName:   commonName,
+			Organization: []string{"juju"},
+			SerialNumber: UUID,
+		},
+		NotBefore:             now.UTC().AddDate(0, 0, -7),
+		NotAfter:              expiry.UTC(),
+		SubjectKeyId:          bigIntHash(key.N),
+		KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		IsCA:                  true,
+		MaxPathLen:            0, // Disallow delegation for now.
+		BasicConstraintsValid: true,
+	}
+	certDER, err := x509.CreateCertificate(rand.Reader, template, template, &key.PublicKey, key)
+	if err != nil {
+		return "", "", fmt.Errorf("cannot create certificate: %v", err)
+	}
+	certPEMData := pem.EncodeToMemory(&pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: certDER,
+	})
+	keyPEMData := pem.EncodeToMemory(&pem.Block{
+		Type:  "RSA PRIVATE KEY",
+		Bytes: x509.MarshalPKCS1PrivateKey(key),
+	})
+	return string(certPEMData), string(keyPEMData), nil
+}
+
+// newSerialNumber returns a new random serial number suitable
+// for use in a certificate.
+func newSerialNumber() (*big.Int, error) {
+	// A serial number can be up to 20 octets in size.
+	// https://tools.ietf.org/html/rfc5280#section-4.1.2.2
+	n, err := rand.Int(rand.Reader, new(big.Int).Lsh(big.NewInt(1), 8*20))
+	if err != nil {
+		return nil, errors.Annotatef(err, "failed to generate serial number")
+	}
+	return n, nil
+}
+
+// NewDefaultServer generates a certificate/key pair suitable for use by a server, with an
```

Comment:
can we just keep these two functions (NewDefaultServer and NewServer) in juju/cert, and expose newLeaf (NewLeaf)?